Fix problem with incorrect spellcheck spans

sbaughman opened the pull request:
Adding the `if` checks in the returns of `WordWalker.getNextWord` caused an issue where if the word was an empty string, the loop would continue and this could cause a returned range to include multiple words.  This caused spellcheck spans to get wrapped around multiple words incorrectly, as well as potentially causing nested spellcheck spans. 

Instead, the responsibility of ignoring empty strings is moved to the functions where `WordWalker.getNextWord` is called.   This was already the case in `markTyposInRange`, and is now added to `getWords`, as well as the `getWordsWithWordWalker` function in the `wordwalker.js` test.

sbaughman committed:
Remove 'if' checks on WordWalker returns

when the if check failed to execute because word was an empty string,
the loop would continue and the ranges would then include multiple words

sbaughman committed:
Only check non-blank words

sbaughman committed:
Only return non-empty words

caffodian commented on plugins/nanospell/plugin.js:
```diff
@@ -190,7 +190,7 @@
 			// so just return what we've walked
 			// of the current word.
-			if (word) return { word: word, range: wordRange };
+			return { word: word, range: wordRange };
```
Is this actually a revert?  Looks familiar.

sbaughman replied:
Yes it is a reversion.  What I was trying to prevent by the `if` check was unnecessarily returning an empty word `""`.  However, if the word currently is a blank word, you don't trigger a return, so it continues through the loop (at least in the case of the top two returns - the bottom return is because you hit the end of the text node anyway, so a change there doesn't make a difference).  Since you are continuing, the text node range offset does not reset so subsequent returns of non-blank words can include spaces and/or multiple words.  This affects words that are sent in the AJAX call, as well as `span` wrapping by `markTyposInRange`, since both have their data supplied by the `WordWalker`

caffodian replied:
This makes sense.  I guess the other option would be that we could bump the offset up before returning.

caffodian replied:
At the cost of much more complexity (have to consume all non-word characters before returning)

sbaughman replied:
In the course of originally answering your question, I realized that solution as well and added it in.  Between that and changing how the end of the range was set at the end of the `for` loop, it seems to be working quite well!

caffodian commented on plugins/nanospell/plugin.js:
```diff
@@ -641,14 +641,16 @@
-					words.push(word.word);
+				while(currentWordObj = wordwalker.getNextWord()) {
+					word = currentWordObj.word;
+					if (word) words.push(word);
```
Why does it work if we pull the if statement out to here?  There is a subtle difference, but it doesn't immediately click.  This would probably be worth some inline commenting.

caffodian replied:
This would be a good place to document that word can be empty, and we just ignore it.

caffodian commented on tests/typowalking/wordwalker.js:
```diff
@@ -27,7 +28,8 @@ bender.test( {
 		while (currWordObj = wordwalker.getNextWord()) {
-			wordsReturned.push(currWordObj.word);
+			word = currWordObj.word;
+			if (word) wordsReturned.push(word);
```
Can you think of any ways to represent the original bug as a test?  I think trying to time it will suck badly, but is there any way to maybe set the html and just trigger the specific walking event?

caffodian replied:
I wouldn't spend too much time on this test though.  If it's not obvious to set up...

sbaughman replied:
there may be some validity to eventually setting up tests to test the word ranges returned, not just the text of the words, as they are separate object properties.  In the course of troubleshooting this issue, I ran across several cases where the returned word range did not correspond exactly to the returned text, which I'm sure caused some of the observed issues.

The separation of words for the AJAX call at the beginning of spellcheck relies on the actual `word` text while the insertion of spellcheck spans relies on the `wordRange` property, so it makes sense to test both.

caffodian replied:
Hmm.

Maybe something like... set the editor html to get the issue (wordwalker stumbles on an empty word) -> invoke wordwalker on the range a few times and see what comes back?

sbaughman committed:
Set start of wordRange to next non-separator offset if a word separator is encountered but there is no valid text in the current word

Avoids returning blank words and corrects start of range for next word

sbaughman committed:
Change how end of wordRange is set

This sets more accurately because the offset is always in the context
of the current text node.  Avoids problems when line breaks and other
such elements are encountered

sbaughman committed:
Only return at end of loop if there is a valid word to return

sbaughman committed:
remove if check from words returned

after modifying word walker, it should no longer ever return blank
words.

sbaughman committed:
Merge pstat_rewrite

sbaughman committed:
Fix merge conflict

sbaughman committed:
Make getWordsWithWordWalker return words and ranges, not just words

sbaughman committed:
Clone contents instead of extracting contents

sbaughman committed:
Add function to return array of range contents.  Rename getWordsWithWordWalker function

sbaughman committed:
Start refactoring tests to use new word objects and make range assertions

sbaughman committed:
Finish refactoring tests

caffodian reviewed (approved):
looks good to me.  The words vs. ranges thing is a bit hard to understand still, but it's also a hard problem
